Validate workflow YAML in CI (yq lint step)

The validate.yml workflow currently has no self-check — invalid workflow YAML only fails when GitHub Actions parses it. Add a step (or a separate workflow) that runs `yq eval '.' .github/workflows/*.yml` to catch syntax errors locally and in CI before they reach GitHub. This is consistent with how validate-schemas.sh checks agent YAML.

_Follow-up from #5_

# Implementation Plan

Now I have a thorough understanding of the existing patterns. Here is the implementation plan:

---

# Implementation Plan: Issue #83 — Validate Workflow YAML in CI (yq lint step)

## 1. Objective

Add a `yq eval '.'` lint step that validates `.github/workflows/*.yml` files for YAML syntax errors, both locally (pre-commit hook) and in CI (GitHub Actions), before GitHub Actions parses them. This mirrors the existing pattern in `validate-schemas.sh` and `check-dangerous-flags.sh`.

---

## 2. Approach

- Create a standalone shell script `scripts/lint-workflow-yaml.sh` following the same structure as `scripts/check-dangerous-flags.sh`: accepts file arguments, defaults to scanning `.github/workflows/`, collects all errors before reporting, exits 1 on failure.
- Add a step to `.github/workflows/validate.yml` that runs the new script on all workflow files. Trigger it on `paths: ['.github/workflows/**']` PRs in addition to the existing `agents/`, `fleets/`, `tests/` triggers.
- Extend `hooks/pre-commit` to also run the lint script on staged `*.yml` files under `.github/workflows/`.
- No new workflow file is needed — the existing `validate.yml` is the right place.

**Key decisions:**
- Use `yq eval '.'` (not `yq validate`) for portability across yq v3 and v4.
- Reuse yq that is already installed in `validate.yml` — no extra install step needed.
- Scope the CI `on.pull_request.paths` to also include `.github/workflows/**` so the step fires when workflow files change.

---

## 3. Files to Create

| File | Description |
|------|-------------|
| `scripts/lint-workflow-yaml.sh` | Standalone lint script. Accepts file list or defaults to `.github/workflows/*.yml`. Runs `yq eval '.' <file>` on each, accumulates errors, prints summary, exits 1 on any failure. |

---

## 4. Files to Modify

| File | Specific Changes |
|------|-----------------|
| `.github/workflows/validate.yml` | (a) Add `.github/workflows/**` to `on.pull_request.paths`. (b) Add a new step `Lint workflow YAML` that runs `./scripts/lint-workflow-yaml.sh` after the schema-validation step. |
| `hooks/pre-commit` | Add a block that collects staged `.github/workflows/*.yml` files and passes them to `scripts/lint-workflow-yaml.sh`, mirroring the existing dangerous-flags block. |

---

## 5. Implementation Order

1. **Create `scripts/lint-workflow-yaml.sh`**
   - Copy structure from `scripts/check-dangerous-flags.sh` (shebang, `set -euo pipefail`, `SCRIPT_DIR`/`REPO_ROOT`, argument handling, loop, error accumulation, summary).
   - Core check: `yq eval '.' "$file" > /dev/null 2>&1 || { ... record error ... }`.
   - Default glob: `"${REPO_ROOT}/.github/workflows"/*.yml` when no args provided.
   - Make executable: `chmod +x scripts/lint-workflow-yaml.sh`.

2. **Modify `.github/workflows/validate.yml`**
   - Extend `on.pull_request.paths` to add `- '.github/workflows/**'`.
   - Add step after the `Validate YAML schemas` step:
     ```yaml
     - name: Lint workflow YAML
       run: ./scripts/lint-workflow-yaml.sh
     ```
   - yq is already installed in a prior step; no additional install needed.

3. **Modify `hooks/pre-commit`**
   - After the existing staged-file collection blocks, add:
     ```bash
     STAGED_WORKFLOWS=$(git diff --cached --name-only --diff-filter=ACM | grep '^\.github/workflows/.*\.yml$' || true)
     if [ -n "$STAGED_WORKFLOWS" ]; then
       # shellcheck disable=SC2086
       "$REPO_ROOT/scripts/lint-workflow-yaml.sh" $STAGED_WORKFLOWS || HOOK_FAILED=1
     fi
     ```
   - Ensure `HOOK_FAILED` exit propagation is consistent with the existing pattern.

4. **Self-validate**: Run `./scripts/lint-workflow-yaml.sh` manually against the existing workflow files to confirm they pass before committing.

---

## 6. Verification

| Test | How |
|------|-----|
| Script exits 0 on valid YAML | `./scripts/lint-workflow-yaml.sh` against current `.github/workflows/*.yml` — must succeed |
| Script exits 1 on invalid YAML | Create a temp file with broken YAML, pass it as argument, confirm non-zero exit and error message |
| Script output format | Verify error lines include file path and a clear message |
| CI step fires on workflow change | Open a PR that modifies a workflow file; confirm `validate.yml` runs the new step |
| CI step fires on agent/fleet change | Open a PR that modifies `agents/`; confirm the workflow step still runs (existing paths still trigger it) |
| Pre-commit hook catches bad workflow YAML | Stage a broken `.github/workflows/test.yml`, run `git commit`, confirm hook blocks it |
| Pre-commit hook passes clean workflow YAML | Stage a valid workflow change, confirm hook does not block it |

---

## 7. Skills Used

**Invoked during planning:**
- `Explore` subagent — explored `.github/workflows/`, `scripts/`, `tests/`, and `hooks/` to understand existing patterns

**Team knowledge base skills referenced (from Prior Learnings):**
- `agent-validate-config` — same yq-based YAML lint-before-CI pattern
- `bare-mojo-call-lint-guard` — confirmed pairing a shell script guard with a GitHub Actions step on `*.yml` glob
- `adr009-ci-pattern-update` — updates CI workflow patterns; informs where to add the step in `validate.yml`
- `doc-validate-markdown` — "validate file syntax before CI catches it" pattern, same motivation
- `yaml-long-line-refactor` — confirms yq usage patterns in this codebase

---
*Generated by Claude Code Planner*

--- a/.actionlint.yaml
+++ b/.actionlint.yaml
@@ -1,0 +1,9 @@
+# actionlint configuration
+# https://github.com/rhysd/actionlint/blob/main/docs/config.md
+
+# Explicitly enable shellcheck integration so actionlint lints the shell
+# scripts inside run: steps (#139). By default actionlint skips shellcheck
+# when the binary is not found; with this config the CI step passes
+# -shellcheck shellcheck on the command line so it is always enabled.
+self-hosted-runner:
+  labels: []

--- a/.github/actions/setup-pixi/action.yml
+++ b/.github/actions/setup-pixi/action.yml
@@ -22,3 +22,11 @@ runs:
           ~/.cache/rattler/cache
         key: pixi-${{ runner.os }}-${{ hashFiles('pixi.lock') }}
         restore-keys: pixi-${{ runner.os }}-
+
+    - name: Add pixi env tools to PATH
+      shell: bash
+      run: |
+        env_bin=".pixi/envs/${{ inputs.environment }}/bin"
+        if [[ -d "$env_bin" ]]; then
+          echo "$(pwd)/${env_bin}" >> "$GITHUB_PATH"
+        fi

--- a/.github/workflows/validate.yml
+++ b/.github/workflows/validate.yml
@@ -51,7 +51,9 @@ jobs:
           actionlint --version
 
       - name: Run actionlint
-        run: actionlint
+        # Pass -shellcheck explicitly so actionlint always lints run: step
+        # scripts through shellcheck even if auto-detection would skip it (#139).
+        run: actionlint -shellcheck shellcheck
 
   security:
     name: Security Scan
@@ -75,6 +77,23 @@ jobs:
       - uses: actions/checkout@v4
 
       - uses: ./.github/actions/setup-pixi
+
+      - name: Check dependencies
+        run: |
+          echo "--- Checking required tools ---"
+          yq --version
+          jq --version
+          bash --version | head -1
+          echo "All required dependencies present."
+
+      - name: Lint workflow YAML files
+        run: |
+          echo "--- Validating workflow YAML files ---"
+          find .github/workflows -name '*.yml' | sort | while read -r f; do
+            echo "  checking: $f"
+            yq eval '.' "$f" > /dev/null
+          done
+          echo "All workflow YAML files are well-formed."
 
       - name: Validate all YAML files
         env:

